### PR TITLE
Allow generating ESM output from npm (non-breaking)

### LINKED ISSUE #1493: Allow generating ES Module (ESM) output from npm
# This is a (Bug Report / Feature Proposal)

## Description
- I'd like to generate an ESM output so that I can leverage top-level await and provisioned concurrency to load secrets during lambda initialization, [as described by this AWS blog post](https://aws.amazon.com/blogs/compute/using-node-js-es-modules-and-top-level-await-in-aws-lambda/)
- Here the relevant part of my `serverless.yml`
```yml
provider:
  name: aws
  runtime: nodejs18.x

plugins:
  - serverless-webpack
  - serverless-offline

package:
  individually: true

custom:
  webpack:
    webpackConfig: 'webpack.config.cjs'
    includeModules: true
```
- My `package.json` includes `"type": "module"`
- And my `webpack.config.cjs` (CJS b/c `serverless-webpack` doesn't support an ESM `webpack.config.js`, but that's a separate issue):
```js
const path = require('path');
const slsw = require('serverless-webpack');
const nodeExternals = require('webpack-node-externals');

const { isLocal } = slsw.lib.webpack;

module.exports = {
  entry: slsw.lib.entries,
  mode: isLocal ? 'development' : 'production',
  target: 'node18',
  externals: [nodeExternals({ importType: 'node-commonjs' })],
  externalsPresets: { node: true },
  externalsType: 'node-commonjs',
  module: {
    rules: [{
      test: /\.ts$/,
      use: [{ loader: 'ts-loader', options: { transpileOnly: true } }],
      exclude: /node_modules/,
    }],
  },
  resolve: {
    extensions: ['.ts', '.js'],
    modules: ['node_modules', path.resolve(__dirname, './src')],
  },
  experiments: { outputModule: true },
  output: {
    library: { type: 'module' },
    path: path.join(__dirname, '.webpack'),
    module: true,
  },
};
```
- This works, but it produces `[name].mjs` outputs, [which breaks `serverless-offline`](https://github.com/serverless/serverless/issues/11308) 
- So I set my webpack `output.filename` to `[name].js` and then local / `serverless-offline` works again!
- But now deployed code fails b/c Lambda doesn't know this is an ESM output
- Specifically, the `package.json` generated by `serverless-webpack` is missing the `"type": "module"` that I set as my project default

## Additional Data

* ***Serverless-Webpack Version you're using***: 5.11.0
* ***Webpack version you're using***: 5.85.1
* ***Serverless Framework Version you're using***: 3.32.2
* ***Operating System***: macOS Ventura 13.3.1

### PR DESCRIPTION
**Pull request by aldenquimby**
## What did you implement:

Closes #1493

## How did you implement it:
- non-breaking alternative approach to #1494
- [see this thread](https://github.com/serverless-heaven/serverless-webpack/pull/1494#discussion_r1232658185) to understand why that was a breaking change

## How can we verify it:
- my project works when I include this change locally
- generating `.js` file allows `serverless invoke local` to work, and including `type: module` allows the deployed Lambda to work:
image

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [ ] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by aldenquimby** [comment]
@vicary @j0k3r what are your thoughts on this PR? Anything I can do to help move it forward?

And apologies if tagging you is not the appropriate next step. I'd be happy to start a `CONTRIBUTING.md` if you let me know the expected process

**Comment by vicary** [comment]
I am more inclined towards #1494 because it's a smaller change, we can discuss more in details there.

File: lib/packagers/yarn.js [hunk 1]
```diff
@@ -21,15 +21,15 @@ class Yarn {
...
-  }
-
   // eslint-disable-next-line lodash/prefer-constant
   static get mustCopyModules() {
     return false;
   }
 
+  static copyPackageSectionNames() {
+    return ['resolutions'];
+  }
```
[the remark this answers is not in the document]
Comment (aldenquimby) [reply lib/packagers/yarn.js:31]: Yes I'll add this for yarn too

File: lib/packagers/npm.js [hunk 2]
```diff
@@ -16,15 +16,15 @@ class NPM {
...
-  }
-
   // eslint-disable-next-line lodash/prefer-constant
   static get mustCopyModules() {
     return true;
   }
 
+  static copyPackageSectionNames(packagerOptions) {
+    return packagerOptions.copyPackageSectionNames || [];
+  }
```
[the remark this answers is not in the document]
Comment (aldenquimby) [reply lib/packagers/npm.js:26]: 👍 done!

File: tests/packExternalModules.test.js [hunk 3]
```diff
@@ -229,6 +238,79 @@ describe('packExternalModules', () => {
...
+      fsExtraMock.copy.mockImplementation((from, to, cb) => cb());
+      readFileSyncStub.mockReturnValueOnce(originalPackageJSON);
+      readFileSyncStub.mockImplementation(() => {
+        throw new Error('Unexpected call to readFileSync');
+      });
+      packagerFactoryMock.get('npm').rebaseLockfile.mockImplementation((pathToPackageRoot, lockfile) => lockfile);
+      packagerFactoryMock.get('npm').getProdDependencies.mockReturnValue(BbPromise.resolve({}));
+      packagerFactoryMock.get('npm').install.mockReturnValue(BbPromise.resolve());
+      packagerFactoryMock.get('npm').prune.mockReturnValue(BbPromise.resolve());
+      packagerFactoryMock.get('npm').runScripts.mockReturnValue(BbPromise.resolve());
```
[the remark this answers is not in the document]
Comment (aldenquimby) [reply tests/packExternalModules.test.js:301]: @vicary I've made all of your suggested changes except this one. Can you confirm you'd like me to add a yarn test here, even though the rest of the file doesn't test yarn (it uses a mocked packager and appears to have picked npm arbitrarily)

[the remark this answers is not in the document]
Comment (vicary) [reply tests/packExternalModules.test.js:301]: You're right, we don't need to test a mocked packager twice. Resolving this one.

**Review by vicary — DISMISSED** [review]

**Comment by aldenquimby** [comment]
Thank you for the help on this @vicary !

File: README.md [hunk 4]
```diff
@@ -406,28 +406,46 @@ you should use any version `<5.5 >=5.7.1` as the versions in-between have some n
...
-| ignoreScripts      | bool   | false                 | Do not execute package.json hook scripts on install |
-| noInstall          | bool   | false                 | Do not run `npm install` (assume install completed) |
-| lockFile           | string | ./package-lock.json   | Relative path to lock file to use                   |
+| Option                  | Type     | Default             | Description                                                         |
+|-------------------------|----------|---------------------|---------------------------------------------------------------------|
+| ignoreScripts           | bool     | false               | Do not execute package.json hook scripts on install                 |
+| noInstall               | bool     | false               | Do not run `npm install` (assume install completed)                 |
+| lockFile                | string   | ./package-lock.json | Relative path to lock file to use                                   |
+| copyPackageSectionNames | string[] | []                  | Entries in your `package.json` to copy to the output `package.json` |
```
Comment (j0k3r) [inline README.md:414]: Just by reading that line, it's not clear how it'll be used. Maybe you can put sth as example? Like
```suggestion
| copyPackageSectionNames | string[] | []                  | Entries in your `package.json` to copy to the output `package.json` (ie: ESM output) |
```

Comment (aldenquimby) [reply README.md:414]: 👍 good point. That should help tie this to the section below explaining ESM output too

**Review by j0k3r — COMMENTED** [review]
This looks good to me, can you just squash all commits?

**Review by j0k3r — APPROVED** [review]

**Pull request merged**